code/game/machinery/computer/law.dm

```diff
@@ -27,20 +41,68 @@
 	return
 
 
-/obj/machinery/computer/aiupload/attackby(obj/item/O as obj, mob/user as mob, params)
-	if(istype(O, /obj/item/aiModule))
-		if(!current)//no AI selected
-			to_chat(user, "<span class='danger'>No AI selected. Please chose a target before proceeding with upload.")
-			return
-		var/turf/T = get_turf(current)
-		if(!atoms_share_level(T, src))
-			to_chat(user, "<span class='danger'>Unable to establish a connection</span>: You're too far away from the target silicon!")
-			return
+/obj/machinery/computer/aiupload/attackby(obj/item/O, mob/user, params)
+	if(!istype(O, /obj/item/aiModule))
+		return
+	if(!current)//no AI selected
+		to_chat(user, "<span class='danger'>No AI selected. Please choose a target before proceeding with upload.</span>")
+		return
+	var/turf/T = get_turf(current)
+	if(!atoms_share_level(T, src))
+		to_chat(user, "<span class='danger'>Unable to establish a connection</span>: You're too far away from the target silicon!</span>")
+		return
+	if(!emagged) //non-emag law change
 		var/obj/item/aiModule/M = O
 		M.install(src)
 		return
+	if(world.time < cooldown)
+		to_chat(user, "<span class='danger'>The program seems to have frozen. It will need some time to process.</span>")
+		return
+	do_sparks(5, TRUE, src)
+	countlaws()
+	if(!emag_ion_check())
+		emag_inherent_law()
 	return ..()
 
+/obj/machinery/computer/aiupload/proc/countlaws()
+	foundlaws = 0
+	for(var/datum/ai_law/law in current.laws.inherent_laws)
+		foundlaws++
+
+/obj/machinery/computer/aiupload/proc/emag_ion_check()
+	var/emag_law = new /datum/ai_law/inherent(generate_ion_law()).law
```

Comment:
We tend to avoid doing `someproc().somevar`, so instead this needs to be something like:
```dm
var/datum/ai_law/inherent/new_law = new(generate_ion_law())
var/emag_law = new_law.law
```
Same with the other warnings below.

Also in this case, I would move this code down to the block where you need it, at the `current.laws.ion_laws[1].law = emag_law` line. That way we're not newing things unnecessarily.